src/hotspot/share/adlc/formssel.cpp

```diff
@@ -3813,51 +3813,77 @@ bool MatchNode::equivalent(FormDict &globals, MatchNode *mNode2) {
     "AndI","AndL",
-    "AndV",
     "MaxI","MinI","MaxF","MinF","MaxD","MinD",
-    "MaxV", "MinV",
     "MulI","MulL","MulF","MulD",
-    "MulVB","MulVS","MulVI","MulVL","MulVF","MulVD",
     "OrI","OrL",
-    "OrV",
-    "XorI","XorL",
-    "XorV"
+    "XorI","XorL"
   };
-  int cnt = sizeof(commut_op_list)/sizeof(char*);
 
-  if( _lChild && _rChild && (_lChild->_lChild || _rChild->_lChild) ) {
+  static const char *commut_vector_op_list[] = {
+    "AddVB", "AddVS", "AddVI", "AddVL", "AddVF", "AddVD",
+    "MulVB", "MulVS", "MulVI", "MulVL", "MulVF", "MulVD",
+    "AndV", "OrV", "XorV",
+    "MaxV", "MinV"
+  };
+
+  if (_lChild && _rChild && (_lChild->_lChild || _rChild->_lChild)) {
     // Don't swap if right operand is an immediate constant.
     bool is_const = false;
-    if( _rChild->_lChild == NULL && _rChild->_rChild == NULL ) {
+    if (_rChild->_lChild == NULL && _rChild->_rChild == NULL) {
       FormDict &globals = _AD.globalNames();
       const Form *form = globals[_rChild->_opType];
-      if ( form ) {
-        OperandForm  *oper = form->is_operand();
-        if( oper && oper->interface_type(globals) == Form::constant_interface )
+      if (form) {
+        OperandForm *oper = form->is_operand();
+        if (oper && oper->interface_type(globals) == Form::constant_interface)
           is_const = true;
       }
     }
-    if( !is_const ) {
-      for( int i=0; i<cnt; i++ ) {
-        if( strcmp(_opType, commut_op_list[i]) == 0 ) {
-          count++;
-          _commutative_id = count; // id should be > 0
+
+    if (!is_const) {
+      int scalar_cnt = sizeof(commut_op_list)/sizeof(char*);
+      int vector_cnt = sizeof(commut_vector_op_list)/sizeof(char*);
+      bool matched = false;
+
+      // Check the commutative vector op first. It's noncommutative if
+      // the current node is a masked vector op, since a mask value
+      // is added to the original vector node's input list and the original
+      // first two inputs are packed into one BinaryNode. So don't swap
+      // if one of the operands is a BinaryNode.
+      for (int i = 0; i < vector_cnt; i++) {
+        if (strcmp(_opType, commut_vector_op_list[i]) == 0) {
+          if (strcmp(_lChild->_opType, "Binary") != 0 &&
+              strcmp(_rChild->_opType, "Binary") != 0) {
+            count++;
+            _commutative_id = count; // id should be > 0
+          }
+          matched = true;
           break;
         }
       }
+
+      // Then check the scalar op if the current op is not in
+      // the commut_vector_op_list.
+      if (!matched) {
+        for (int i = 0; i < scalar_cnt; i++) {
+          if (strcmp(_opType, commut_op_list[i]) == 0) {
+            count++;
+            _commutative_id = count; // id should be > 0
+            break;
+          }
+        }
+      }
     }
   }
-  if( _lChild )
+  if (_lChild)
     _lChild->count_commutative_op(count);
-  if( _rChild )
+  if (_rChild)
     _rChild->count_commutative_op(count);
 }
```

Comment:
Matcher rules based on proposed masked IR shall look like following
   match(Set dst  AddVI (Binary vsrc1 vsrc2)  mask)

AddVI is still commutative and generated DFA will check for appropriate child states i.e.
    kid[0]->state == _Binary_vec_vec  &&  kid[1]->state == _mask

So even if matcher generates additional check by swapping the states of the two child nodes it should still be ok. Can you kindly elaborate the need for this change.

Reply:
Yes, actually that's ok if matcher swaps the two operands. Avoiding the swapping could avoid generating unneeded rules which could reduce the whole code size.